components/camel-pubnub/src/main/java/org/apache/camel/component/pubnub/PubNubProducer.java

```diff
@@ -0,0 +1,261 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.pubnub;
+
+import java.util.Arrays;
+
+import com.pubnub.api.PubNubException;
+import com.pubnub.api.callbacks.PNCallback;
+import com.pubnub.api.models.consumer.PNPublishResult;
+import com.pubnub.api.models.consumer.PNStatus;
+import com.pubnub.api.models.consumer.history.PNHistoryResult;
+import com.pubnub.api.models.consumer.presence.PNGetStateResult;
+import com.pubnub.api.models.consumer.presence.PNHereNowResult;
+import com.pubnub.api.models.consumer.presence.PNSetStateResult;
+import com.pubnub.api.models.consumer.presence.PNWhereNowResult;
+
+import org.apache.camel.AsyncCallback;
+import org.apache.camel.CamelException;
+import org.apache.camel.Exchange;
+import org.apache.camel.impl.DefaultAsyncProducer;
+import org.apache.camel.util.ObjectHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The PubNub producer.
+ */
+public class PubNubProducer extends DefaultAsyncProducer {
+    private static final Logger LOG = LoggerFactory.getLogger(PubNubProducer.class);
+    private final PubNubEndpoint endpoint;
+    private final PubNubConfiguration pubnubConfiguration;
+
+    public PubNubProducer(PubNubEndpoint endpoint, PubNubConfiguration pubNubConfiguration) {
+        super(endpoint);
+        this.endpoint = endpoint;
+        this.pubnubConfiguration = pubNubConfiguration;
+    }
+
+    @Override
+    public boolean process(final Exchange exchange, final AsyncCallback callback) {
+
+        Operation operation = getOperation(exchange);
+
+        LOG.debug("Executing {} operation", operation);
+
+        switch (operation) {
+        case PUBLISH: {
+            doPublish(exchange, callback);
+            break;
+        }
+        case FIRE: {
+            doFire(exchange, callback);
+            break;
+        }
+        case GETHISTORY: {
+            doGetHistory(exchange, callback);
+            break;
+        }
+        case GETSTATE: {
+            doGetState(exchange, callback);
+            break;
+        }
+        case HERENOW: {
+            doHereNow(exchange, callback);
+            break;
+        }
+        case SETSTATE: {
+            doSetState(exchange, callback);
+            break;
+        }
+        case WHERENOW: {
+            doWhereNow(exchange, callback);
+            break;
+        }
+        default:
+            throw new UnsupportedOperationException(operation.toString());
+        }
+        if (exchange.getException() != null) {
```

Comment:
This code should ideally be removed as you do this after you have send the message, instead its maybe a try .. catch you may need in case an exception was thrown before sending the message, which you then need to catch and call callback.done(true)